blog/2019-09-12-audit-logs-with-change-data-capture-and-kafka-streams.adoc

```diff
@@ -0,0 +1,656 @@
+            .getSingleResult();
+        String useCase = ctx.getMethod().getAnnotation(Audited.class).useCase();
+
+        TransactionContextData context = new TransactionContextData(); // <5>
+
+        context.transactionId = txtId.longValueExact();
+        context.userName = jwt.<String>claim("sub").orElse("anonymous");
+        context.clientDate = getRequestDate();
+        context.useCase = useCase;
+
+        entityManager.persist(context);
+
+        return ctx.proceed(); // <6>
+    }
+
+    private ZonedDateTime getRequestDate() {
+        String requestDate = request.getHeader(HttpHeaders.DATE);
+        return requestDate != null ?
+            ZonedDateTime.parse(requestDate, DateTimeFormatter.RFC_1123_DATE_TIME) :
+            null;
+    }
+}
+----
+<1> `@Interceptor` and `@Audited` mark this as an interceptor bound to our custom `@Audited` annotion.
+<2> The `@Priority` annotation controls at which point in the interceptor stack the auditing interceptor should be invoked.
+Any application-provided interceptors should have a priority larger than `Priority.APPLICATION` (2000);
+in particular, this ensures that a transaction will have been started before by means of the `@Transactional` annotation and its accompanying interceptor which run in the `Priority.PLATFORM_BEFORE` range (< 1000).
+<3> The caller's JWT token injected via the https://microprofile.io/project/eclipse/microprofile-jwt-auth[MicroProfile JWT RBAC] API
+
+For each audited method the interceptor fires and will
+
+* obtain the current transaction id (the exact way for doing so is database-specific, in the example the `txid_current()` function from Postgres is called) pass:[<i class="conum" data-value="4"></i>]
+* persist a `TransactionContextData` entity via JPA; its primary key value is the transaction id selected before, and it has attributes for the user name (obtained from the JWT token),
+the request date (obtained from the `DATE` HTTP request header) and the use case identifier (obtained from the `@Audited` annotation of the invoked method) pass:[<i class="conum" data-value="5"></i>]
+* continue the call flow of the invoked method pass:[<i class="conum" data-value="6"></i>]
+
+When invoking the REST service to create and update a few vegetables,
+the following records should be created in the database
+(refer to the README in the provided example for instructions on building the example code and https://github.com/debezium/debezium-examples/tree/master/auditlog#inserting-some-data-and-observing-the-audit-log[invoking the vegetable service] with a suitable JWT token):
+
+[source,sql]
+----
+vegetablesdb> select * from inventory.vegetable;
++------+---------------+---------+
+| id   | description   | name    |
+|------+---------------+---------|
+| 1    | Spicy         | Potatoe |
+| 11   | Delicious!    | Banana  |
+| 10   | tasty         | Pear    |
++------+---------------+---------+
+----
+
+[source,sql]
+----
```

Comment:
Minor nit pick, why in the example above in the trasactions table the usernames are with space and below without?

Reply:
Good catch; Updating the image.